pkg/driver/helvarnet/health.go

```diff
@@ -0,0 +1,165 @@
+	{"BatteryFail", "Battery has failed", 0x00040000},
+	{"Inhibit", "Prevents an emergency fitting from going into emergency mode", 0x00200000},
+	{"FTRequested", "Emergency Function Test has been requested", 0x00400000},
+	{"DTRequested", "Emergency Duration Test has been requested", 0x00800000},
+	{"Unknown", "Initial state of an emergency fitting", 0x01000000},
+	{"OverTemperature", "Load is over temperature/heating", 0x02000000},
+	{"OverCurrent", "Too much current is being drawn by the load", 0x04000000},
+	{"CommsError", "Communications error", 0x08000000},
+	{"SevereError", "Indicates that a load is either over temperature or drawing too much current, or both", 0x10000000},
+	{"BadReply", "Indicates that a reply to a query was malformed", 0x20000000},
+	{"DeviceMismatch", "The actual load type does not match the expected type", 0x80000000},
+}
+
+// GetStatusListFromFlag takes a bitwise flag integer and returns a list of State structs
+// corresponding to the set bits in the flag.
+func GetStatusListFromFlag(flag int64) []State {
+
+	if flag == 0 {
+		return []State{}
+	}
+
+	var statusList []State
+	for _, ds := range DeviceStatuses {
+		if flag&ds.FlagValue != 0 {
+			statusList = append(statusList, ds)
+		}
+	}
+
+	return statusList
+}
+
+func updateDeviceFaults(ctx context.Context, status int64, fc *healthpb.FaultCheck, raisedFaults map[int64]bool) {
+
+	// Handle negative status codes for special conditions, mainly comms issues
+	if status < 0 {
+		rel := &gen.HealthCheck_Reliability{
+			UnreliableTime: timestamppb.Now(),
+		}
+		switch status {
+		case DeviceOfflineCode:
+			rel.State = gen.HealthCheck_Reliability_NO_RESPONSE
+			fc.SetFault(&gen.HealthCheck_Error{
+				SummaryText: "Device Offline",
+				DetailsText: "No communication received from device since the last Smart Core restart",
+				Code: &gen.HealthCheck_Error_Code{
+					Code:   strconv.Itoa(DeviceOfflineCode),
+					System: SystemName,
+				},
+			})
+		case BadResponseCode:
+			rel.State = gen.HealthCheck_Reliability_BAD_RESPONSE
+			fc.SetFault(&gen.HealthCheck_Error{
+				SummaryText: "Bad Response",
+				DetailsText: "The device has sent an invalid response to a command",
+				Code: &gen.HealthCheck_Error_Code{
+					Code:   strconv.Itoa(BadResponseCode),
+					System: SystemName,
+				},
+			})
+		default:
+			// this should really never happen, but if it does, then it is a problem with the driver
+			// and it should be flagged
+			rel.State = gen.HealthCheck_Reliability_UNRELIABLE
+			fc.SetFault(&gen.HealthCheck_Error{
+				SummaryText: "Internal Driver Error",
+				DetailsText: "The device has an unrecognised internal status code",
+				Code: &gen.HealthCheck_Error_Code{
+					Code:   strconv.Itoa(UnrecognisedErrorCode),
+					System: SystemName,
+				},
+			})
+		}
+		fc.UpdateReliability(ctx, rel)
```

Comment:
Setting a fault will implicitly make the device reliable, which you're then potentially resetting again here. If you want to treat successful responses that happen to contain errors like Offline as reliability issues then you shouldn't set the fault and just update the reliability. The Reliability type has a last_error field that looks the same as fault, so you can use that to capture codes and so on.